### ISSUE TITLE
Add Parametric [Intrinsic + Extrinsic] Calibration.  

### ISSUE THREAD

**Issue by RameezI** [RameezI/TUeLaneTracker#2]
The Base and Purview Histograms bins should be correctly calculated by the algorithm for any new camera source, given the right basic parameters. One idea, currently under investigation, is to calculate pixel-to-cm ratio for each image line. @TvdSmagt

**Comment by TvdSmagt** [comment]
Should I add this in the Camera.h file?

**Comment by RameezI** [comment]
@TvdSmagt 
The Camera.h file already has a `getCM_TO_PIXEL()` method which calculates the cm to pixel ratio of the last row in the image. However, this method is not accurate and does not scale well. I  have added  a skeleton method `getCameraMatrix()` which is a more generic way of describing camera and pose parameters. This code is still in the dev branch. 

The calibration matrices will be computed by a separate program, added a 'CameraCallibrator'  executable, which will write the computed matrices to a .yaml file.

The challenges are:
1. Obtain extrinsic and intrinsic camera matrices. [code should go in main.cpp of CameraCallibrator]
2. Replace usage of `getCM_TO_PIXEL()` with that of calibration matrices. [This will mainly effect LaneFilter.cpp] 

Note: At the moment the usage of `getCM_TO_PIXEL()` is scattered in multiple files and I am working on confining it to LaneFilter.cpp file only so you can easily upgrade.

**Comment by TvdSmagt** [comment]
Currently, I have worked on adding the intrinsic and extrinsic matrix calculations based on known parameters to the CameraCalibrator file.

As for the cm to pixel ratio, it is dependent on the used horizon. If this horizon is set higher, then the pixel to cm ratio will grow slower.

The formula's I have derived to calculate the width of all pixels (up to the horizon) are as follows:

Z0 = ty * tan(90 - FOVy / 2)

where Z0 is the distance of the bottom line of pixels to the camera, ty is the vertical translation (or height) of the camera, and FOVy is the vertical field of view of the camera. With this distance, we can calculate the width of the bottom line as:

W0 = 2 * Z0 * tan(FOVx /2)

Where W0 is the width of the bottom line of pixels and FOVx is the horizontal field of view. This relation will hold for a variable Z, dependent on the pixel line height:

Zx = ty * tan(90 - theta_y)

where Zx is the longitudinal distance from the camera for pixel line angle theta_y. This can then be used to calculate the total width of the pixel line in (c)m:

Wx = 2*Zx * tan(theta_y)

The theta_y used here can be calculated as:

theta_y = (Py-y/2)/(y/2) * FOVy / 2

where Py is the vertical pixel location (assuming top is pixel row 1 and bottom is the pixel row with the highest number), and y is the vertical pixel size.

If the horizon is not exactly in the middle, then this angle will have to be added to the argument within each tan statement.
